File: java/scon-core/src/main/java/io/github/realmlabs/scon/Ast.java
package io.github.realmlabs.scon;

import java.util.List;

final class Ast {
    record Document(ObjectNode root, String file) {}
    record ObjectNode(List<Member> members, Span span) {}
    sealed interface Member permits Field, Include, ObjectSpread { Span span(); }
    record Field(PathNode path, ValueNode value, Span span) implements Member {}
    record Include(StringNode path, Span span) implements Member {}
    record ObjectSpread(SubstitutionNode sub, Span span) implements Member {}
    record PathNode(List<PathSegment> segments, Span span) {}
    record PathSegment(String value, boolean quoted, Span span) {}
    sealed interface ValueNode permits NullNode, BoolNode, NumberNode, StringNode, ArrayNode, ObjectValueNode, SubstitutionNode {
        Span span();
    }
    record NullNode(Span span) implements ValueNode {}
    record BoolNode(boolean value, Span span) implements ValueNode {}
    record NumberNode(String raw, Span span) implements ValueNode {}
    record StringNode(String value, String raw, List<StringPart> parts, Span span) implements ValueNode {}
    record ArrayNode(List<ArrayItem> items, Span span) implements ValueNode {}
    record ObjectValueNode(ObjectNode object, Span span) implements ValueNode {}
    record SubstitutionNode(PathNode path, Span span) implements ValueNode {}
    sealed interface ArrayItem permits ArrayValue, ArraySpread { Span span(); }
    record ArrayValue(ValueNode value, Span span) implements ArrayItem {}
    record ArraySpread(SubstitutionNode sub, Span span) implements ArrayItem {}
    sealed interface StringPart permits StringLiteral, StringInterpolation {}
    record StringLiteral(String value) implements StringPart {}
    record StringInterpolation(PathNode path, Span span) implements StringPart {}

    private Ast() {}
}

File: kotlin/scon-core/src/main/kotlin/io/github/realmlabs/scon/Ast.kt
package io.github.realmlabs.scon

import java.nio.file.Path

public data class SourceSpan(
    val start: Int,
    val end: Int,
) {
    init {
        require(start <= end) { "source span start must be <= end" }
    }
}

public data class ParsedDocument(
    val sourceName: String,
    val sourcePath: Path?,
    val root: AstObject,
    val tokens: List<SconToken>,
)

public sealed interface AstNode {
    public val span: SourceSpan
}

public data class AstObject(
    val members: List<AstObjectMember>,
    override val span: SourceSpan,
    val explicitBraces: Boolean,
) : AstNode

public sealed interface AstObjectMember : AstNode

public data class AstField(
    val path: AstPath,
    val value: AstValue,
    override val span: SourceSpan,
) : AstObjectMember

public data class AstInclude(
    val path: AstString,
    override val span: SourceSpan,
) : AstObjectMember

public data class AstObjectSpread(
    val substitution: AstSubstitution,
    override val span: SourceSpan,
) : AstObjectMember

public data class AstPath(
    val segments: List<AstPathSegment>,
    override val span: SourceSpan,
) : AstNode

public data class AstPathSegment(
    val value: String,
    val quoted: Boolean,
    override val span: SourceSpan,
) : AstNode

public sealed interface AstValue : AstNode

public data class AstNull(override val span: SourceSpan) : AstValue
public data class AstBool(val value: Boolean, override val span: SourceSpan) : AstValue
public data class AstNumber(val raw: String, override val span: SourceSpan) : AstValue
public data class AstString(
    val value: String,
    val raw: String,
    val parts: List<AstStringPart>,
    override val span: SourceSpan,
) : AstValue
public data class AstArray(val items: List<AstArrayItem>, override val span: SourceSpan) : AstValue
public data class AstSubstitution(val path: AstPath, override val span: SourceSpan) : AstValue

public data class AstObjectValue(val value: AstObject) : AstValue {
    override val span: SourceSpan = value.span
}

public sealed interface AstArrayItem : AstNode

public data class AstArrayValueItem(
    val value: AstValue,
    override val span: SourceSpan,
) : AstArrayItem

public data class AstArraySpread(
    val substitution: AstSubstitution,
    override val span: SourceSpan,
) : AstArrayItem

public sealed interface AstStringPart

public data class AstStringLiteralPart(val value: String) : AstStringPart

public data class AstStringInterpolationPart(
    val path: AstPath,
    val span: SourceSpan,
) : AstStringPart
